### PR TITLE
Defines for shader and preprocessor

### PR DESCRIPTION
**Pull request by amakaseev**
Added shader defines as optional parameter for use custom defines in programs.
In the future I think to add custom preprocessing.

### CODE REVIEW

**Comment by ivanpopelyshev** [comment]
Good for initial draft :)

**Comment by amakaseev** [comment]
Example usage:

  if (!this._shader) {
    let defines = {};

    if (this._texture) {
      defines.TEXTURE = '';
    }

    if (this._alphaCutoff) {
      defines.ALPHA_CUTOFF = '';
    }

    if (this._lights) {
      defines.LIGHTS = '';
    }

    this._shader = new PIXI.Shader(PIXI.Program.from(defaultVertex, defaultFragment, 'default', 
{defines: defines}));
  }

[default.txt](https://github.com/pixijs/pixi.js/files/4134162/default.txt)

File: packages/core/src/shader/utils/Preprocessor.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
+/**
+ * Created by hazed on 30.01.2020.
+ */
```
Comment (cursedcoder) [inline packages/core/src/shader/utils/Preprocessor.ts]: Is it needed?

File: packages/core/src/shader/utils/Preprocessor.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            else
+            {
+                nameCache[name] = 1;
+            }
+
+            // Setup default define for shader name
+            options.defines.SHADER_NAME = name;
+
+            // Setup custom defines
+            for (const define in options.defines)
```
Comment (cursedcoder) [inline packages/core/src/shader/utils/Preprocessor.ts]: I believe you can use `for (const value of options.defines)`

File: packages/core/src/shader/utils/Preprocessor.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            {
+                const value = options.defines[define];
+
+                this.vertex = `#define ${define} ${value}\n${this.vertex}`;
+                this.fragment = `#define ${define} ${value}\n${this.fragment}`;
```
Comment (cursedcoder) [inline packages/core/src/shader/utils/Preprocessor.ts]: Will it overwrite it on every iteration?

File: packages/core/src/shader/utils/Preprocessor.ts (hunk 1 above)
Comment (bigtimebuddy) [reply packages/core/src/shader/utils/Preprocessor.ts]: Agreed, please remove. We use git history to determine authorship and time stamping.

File: packages/core/src/shader/utils/Preprocessor.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        options.isRawShader = options.isRawShader || (this.vertex.substring(0, 8) === '#version');
+
+        if (!options.isRawShader)
+        {
+            name = name.replace(/\s+/g, '-');
+
+            if (nameCache[name])
+            {
+                nameCache[name]++;
+                // TODO: incremented name in shader code breaks ProgramCache functionality
```
Comment (GoodBoyDigital) [inline packages/core/src/shader/utils/Preprocessor.ts]: would be good to resolve this todo!

File: packages/core/src/shader/utils/Preprocessor.ts (hunk 1 above)
Comment (amakaseev) [reply packages/core/src/shader/utils/Preprocessor.ts]: ok, its auto-generate by WebStorm

File: packages/core/src/shader/utils/Preprocessor.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            options.defines.SHADER_NAME = name;
+
+            // Setup custom defines
+            for (const define in options.defines)
+            {
+                const value = options.defines[define];
+
+                this.vertex = `#define ${define} ${value}\n${this.vertex}`;
+                this.fragment = `#define ${define} ${value}\n${this.fragment}`;
+            }
```
Comment (GoodBoyDigital) [inline packages/core/src/shader/utils/Preprocessor.ts]: this is cool, i think a neater (more performant) way would be to generate the defines in an array and concat them to a string:

```
const defineStrings = [];

for (const define in options.defines)
{
    defineStrings.push(`#define ${define}`);
}

const finalDefines =  defineStrings.join(`\n`);
this.vertex = finalDefines + this.vertex;
this. fragment = finalDefines + this. fragment;

```

File: packages/core/src/shader/utils/Preprocessor.ts (hunk 2 above)
Comment (amakaseev) [reply packages/core/src/shader/utils/Preprocessor.ts]: no, defines must have key and value, as example:
```
#define MAX_BONES 64
uniform mat4 boneMatrices[MAX_BONES];
```

**Pull request closed without merging**